Add iptables and ipsets reconcilers

pippolo84 opened the pull request.

pippolo84 committed:
iptables: Manage rules through runtime reconciliation

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Remove usage of check pkg in tests

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add test for reconciliation loop

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Extract runnable interface from iptablesInterface

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add tests for ipset list

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki committed:
iptables: Use a ticker to decide when to reconcile

The rate limiting effectively limited how quickly the ipset etc. changes were
processed. Use a ticker instead to set the rate at which the changed desired
state is reconciled. Additionally move the retrying to be done via the ticker
as well so new changes to desired state are taken into account when retrying.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
ipset: Add a cell to manage ipsets

Based on the generic reconciler

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Add ip{v4,v6} native routing CIDR

Manage ipv4 and ipv6 native routing CIDRs in the local node struct. This
allows the iptables reconciler to subscribe to the local node update
and:

- get the native routing CIDRs set at startup as a config option
- get the updated v4 native routing CIDRs in case of auto detection
  (when IPAM is one of CRD, ENI, Azure or Alibaba) at runtime

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/iptables/ipset/ipset.go:
```diff
@@ -0,0 +1,269 @@
+			Status: reconciler.StatusPending(),
+		}
+	} else {
+		obj.Addrs[addr] = sets.Empty{}
```
This mutates the object. You must first make a copy. Shallow copy is enough here plus clone of `Addrs`. You can then use the `Insert` method on sets.Set rather than treating it as a map.

joamaki commented on pkg/datapath/iptables/ipset/ipset.go:
```diff
@@ -0,0 +1,269 @@
+	if !found || !obj.Addrs.Has(addr) {
+		return
+	}
+	delete(obj.Addrs, addr)
```
Same here, you cannot mutate the objects returned by queries.

joamaki replied:
Unless of course it's returned by value and not be reference (e.g. `Ipset` and not `*Ipset`, though fields themselves need still a copy)

pippolo84 replied:
I kept the `*Ipset` type since `reconciler.New` requires a `comparable`.

joamaki commented on pkg/datapath/tables/ipset.go:
```diff
@@ -0,0 +1,72 @@
+	return &Ipset{
+		Name:   s.Name,
+		Family: s.Family,
+		Addrs:  s.Addrs.Clone(),
```
No need to clone `Addrs` as we're not mutating it.

joamaki commented on pkg/datapath/tables/ipset.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func (*Ipset) TableHeader() []string {
+	return []string{"Name", "Family", "Addrs"}
```
This is missing "Status"

joamaki commented on pkg/datapath/tables/ipset.go:
```diff
@@ -0,0 +1,72 @@
+	tbl, err := statedb.NewTable[*Ipset](
+		IpsetTableName,
+		IpsetNameIndex,
+		IpsetStatusIndex,
```
No need for status index.

joamaki commented on pkg/datapath/iptables/reconciler.go:
```diff
@@ -0,0 +1,203 @@
+			}
+			state.localNodeInfo = localNodeInfo
+			stateChanged = true
+		case proxyInfo, ok := <-params.proxies:
```
Shouldn't these channels be drained as well as otherwise the methods sending to these could block if the order is unfortunate when stopping?

joamaki committed:
container: Add ImmSet[T], a slice-backed immutable set

Add an immutable set data structure for use in objects stored in
StateDB tables. The ImmSet[T] is useful when storing a relatively
small set of items (1-1000). Implemented as a sorted slice.
Mutations of the set clone the sliceand lookups are implemented
as binary searches.

The benefit of this is a packed presentation, especially for value
types (e.g. netip.Addr) which will cause less overhead for GC.
Downside is that inserts and deletes become expensive when there
are many items.

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki committed:
ipset: Use ImmSet for ipsets

This allows storing the IP address sets in a more compact GC
friendly form and avoids many allocations coming from cloning
of sets.Set (map).

Signed-off-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Avoid useless removals before pruning the entire chain

When removing old Cilium rules, we should not care about rules belonging
to "old" chains, that is, chains prefixes with the "oldCiliumPrefix".
That's because those chains will be entirely removed immediately after
anyway.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add proxy rules incremental reconciliation

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Fix a race in reconciliation loop test

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Add no track rules incremental reconciliation

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Wait for reconciliation for no track pods and proxy updates

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
iptables: Lower reconciliation interval to 50ms

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/iptables/ipset/cell.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/cilium/cilium/pkg/time"
+)
+
+var Cell = cell.Module(
```
Please add a more long-form Go comment to this Cell that describes the different things happening in this cell. E.g. the general idea is that we'd always have a "explain like I'm five" type of comment for each `var Cell` that has enough detail to get a high-level picture of what this is for.

pippolo84 replied:
Addressed in https://github.com/cilium/cilium/pull/31099

joamaki commented on pkg/datapath/iptables/cell.go:
```diff
@@ -16,17 +16,18 @@ var Cell = cell.Module(
 	"iptables",
 	"Handle iptables-related configuration for Cilium",
 
+	ipset.Cell,
```
Let's add a high-level comment for the `var Cell` here as well. Something that summarizes the different things happening here (e.g. mentioning the `CILIUM_` chains, ipset, proxy rules etc.).

pippolo84 replied:
Addressed in https://github.com/cilium/cilium/pull/31099
